### DISCUSSION TITLE
Rule to select specific CAN IDs

### DISCUSSION THREAD

**Discussion by loki429** [lf-edge/ekuiper#3291] [General]
Hi All

I'd like to create a rule that filters on a small set of CAN bus IDs, but without success.

First off I got data type errors:
eg " Where error: between operator cannot compare uint32(1633) and int(512)"

This is the query:
SELECT *, meta(id) AS canId
FROM can0
WHERE canId between 512 AND 519

I've tried using CAST but get no results coming through:
SELECT *, meta(id) AS canId
FROM can0
WHERE CAST(meta(id),INT) BETWEEN 1632 AND 1639;

The rule runs and I see records on source_can0_0 in and out, but op_2_filter shows a matching number in but zero out.
The source data definitely contains data with those CAN ids and I've even tried the rule using values 1 to 99999.

Can anyone see my error and point me in the right direction please?

Thanks,
Jon

![image](https://github.com/user-attachments/assets/e23058d6-5cd2-4446-873f-f8c5840f716c)

**Comment by ngjaying** [comment]
Hi @loki429 Maybe you can print out CanId and see its values. Check this doc about how to debug https://ekuiper.org/docs/en/latest/getting_started/debug_rules.html

**Reply by loki429** [reply]
Thank you for your quick reply.
I've been looking through the debug logs and it looks like a data type problem and working out the correct SQL to operate on it.

With a basic SELECT * query I can see the data and print the canId values.
If I add WHERE I get the following error:
time="2024-10-15 04:53:12" level=error msg="Operation 2_filter error: run Where error: invalid operation uint32(6) > int64(0)" file="node/operations.go:148" rule="can 20x"
time="2024-10-15 04:53:12" level=error msg="Operation 2_filter error: run Where error: invalid operation uint32(1639) > int64(0)" file="node/operations.go:148" rule="can 20x"
time="2024-10-15 04:53:12" level=error msg="Operation 2_filter error: run Where error: invalid operation uint32(1) > int64(0)" file="node/operations.go:148" rule="can 20x"
time="2024-10-15 04:53:12" level=error msg="Operation 2_filter error: run Where error: invalid operation uint32(1638) > int64(0)" file="node/operations.go:148" rule="can 20x"
time="2024-10-15 04:53:12" level=error msg="Operation 2_filter error: run Where error: invalid operation uint32(4) > int64(0)" file="node/operations.go:148" rule="can 20x"
time="2024-10-15 04:53:12" level=error msg="Operation 2_filter error: run Where error: invalid operation uint32(1633) > int64(0)" file="node/operations.go:148" rule="can 20x"
Using:
SELECT *, meta(id) AS canId
FROM can0
WHERE meta(id) > 0

Why is meta(id) of type uint32? 
How can I perform a WHERE on the uint32 or change it to INT? None of the SQL I've been trying is working (eg CAST as INT).
I've tried using WHERE canId > 0 and get the same error.

Thanks,
Jon

**Reply by ngjaying** [reply]
Hi Jon,  the can id was parsed as uint32 internally. Use cast should be fine. Cast only supports The supported types include: bigint, float, string, boolean, bytea and datetime. CAST(meta(id),"bigint") may work.

**Reply by loki429** [reply]
Hello
I've been trying all combinations I can think of overnight but with no success. Unfortunately 'cast' doesn't seem to work.

Surely what I'm trying to achieve is a very basic filter that anyone who uses ek-can has done previously?
I simply want to process data from a few specific CAN bus ids only.

Regarding the cast function:
A very basic query such as the following results in no canId at all.
eg
SELECT *, CAST(meta(id), BIGINT) AS canId
FROM can0

If the canid is parsed internally as uint32 how can we perform any queries on it?
Is the 'cast' function broken?
Is there any alternate way I can filter on the CAN bus ID?

Many thanks,
Jon

**Reply by ngjaying** [reply]
Hi Jon,  I have checked this type conversion problem, it had fixed in the master branch.  I am sorry the ek-can module is not maintained now. NeuronEX (https://docs.emqx.com/en/neuronex/latest/streaming-processing/can.html) has integrated can module. And it is continuously maintained. [Contact us](https://www.emqx.com/en/contact?utm_source=emqx.io&utm_medium=referral&utm_campaign=emqxio-floating-to-contact) if you need more info.

**Reply by loki429** [reply]
Thank you so much for checking on that, much appreciated. I will contact emqx and see if there is a suitable NeuronEX option and price for this use case.
